web/src/main/java/com/microsoft/applicationinsights/web/internal/httputils/HttpServerHandler.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.commons.lang3.Validate;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.http.annotation.Experimental;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.MalformedURLException;
+import java.util.Date;
+
+/**
+ * This Helper Handler class provides the required methods to instrument requests.
+ * @param <P> The HttpRequest entity
+ * @param <Q> The HttpResponse entity
+ */
+@Experimental
+public final class HttpServerHandler<P /* >>> extends @NonNull Object */, Q> {
+
+    /**
+     * Extractor to extract data from request and response
+     */
+    private final HttpExtractor<P, Q> extractor;
+
+    /**
+     * Container that holds collection of
+     * {@link com.microsoft.applicationinsights.web.extensibility.modules.WebTelemetryModule}
+     */
+    private final WebModulesContainer<P, Q> webModulesContainer;
+
+    /**
+     * An instance of {@link TelemetryClient} responsible to track exceptions
+     */
+    private final TelemetryClient telemetryClient;
+
+    /**
+     * Creates a new instance of {@link HttpServerHandler}
+     *
+     * @param extractor The {@code HttpExtractor} used to extract information from request and repsonse
+     * @param webModulesContainer The {@code WebModulesContainer} used to hold
+     *        {@link com.microsoft.applicationinsights.web.extensibility.modules.WebTelemetryModule}
+     * @param telemetryClient The {@code TelemetryClient} used to send telemetry
+     */
+    public HttpServerHandler(HttpExtractor<P, Q> extractor,
+        WebModulesContainer<P, Q> webModulesContainer,
+        /* Nullable */ TelemetryClient telemetryClient) {
+        Validate.notNull(extractor, "extractor");
+        Validate.notNull(webModulesContainer, "WebModuleContainer");
+        this.extractor = extractor;
+        this.webModulesContainer = webModulesContainer;
+        this.telemetryClient = telemetryClient;
+    }
+
+    /**
+     * This method is used to instrument incoming request and initiate correlation with help of
+     * {@link com.microsoft.applicationinsights.web.extensibility.modules.WebRequestTrackingTelemetryModule#onBeginRequest(HttpServletRequest, HttpServletResponse)}
+     * @param request incoming Request
+     * @param response Response object
+     * @return {@link RequestTelemetryContext} that contains correlation information and metadata about request
+     * @throws MalformedURLException
+     */
+    public RequestTelemetryContext handleStart(P request, Q response) throws MalformedURLException {
+        RequestTelemetryContext context = new RequestTelemetryContext(new Date().getTime(),null);
+        RequestTelemetry requestTelemetry = context.getHttpRequestTelemetry();
+        ThreadContext.setRequestTelemetryContext(context);
+
+        String method = extractor.getMethod(request);
+        String userAgent = extractor.getUserAgent(request);
+
```

Comment:
Curious, why do we need a blank space here?

Reply:
It's not needed per say. But when I added that my thought was it keeps nice separation of url related data in a block which is easy to ready. (though I can understand where you are coming from :) )

Reply:
I removed them.